WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift

```diff
@@ -95,10 +95,72 @@ final class MainTabBarControllerTests: XCTestCase {
+        let pushNotificationsManager = MockPushNotificationsManager()
+        ServiceLocator.setPushNotesManager(pushNotificationsManager)
+
+        let storesManager = MockStoresManager(sessionManager: .testingInstance)
+        // Reset `receivedActions`
+        storesManager.reset()
+        ServiceLocator.setStores(storesManager)
+
+        // Trigger `viewDidLoad`
+        XCTAssertNotNil(tabBarController.view)
+        stores.updateDefaultStore(storeID: 134)
+
+        // Simulate successful state resetting after logging out from push notification store switching
+        storesManager.whenReceivingAction(ofType: StatsActionV4.self) { action in
+            if case let .resetStoredStats(completion) = action {
+                completion()
+            }
+        }
+        storesManager.whenReceivingAction(ofType: OrderAction.self) { action in
+            if case let .resetStoredOrders(completion) = action {
+                completion()
+            }
+        }
+        storesManager.whenReceivingAction(ofType: ProductReviewAction.self) { action in
+            if case let .resetStoredProductReviews(completion) = action {
+                completion()
+            }
+        }
+
+        // Action
+        // Send push notification in inactive state
+        let pushNotification = PushNotification(noteID: 1_234, kind: .comment, message: "")
+        pushNotificationsManager.sendInactiveNotification(pushNotification)
+
+        // Simulate that the network call returns a parcel
+        let receivedAction = try XCTUnwrap(storesManager.receivedActions.first as? ProductReviewAction)
+        guard case .retrieveProductReviewFromNote(_, let completion) = receivedAction else {
+            return XCTFail("Expected retrieveProductReviewFromNote action.")
+        }
+        completion(.success(MockProductReviewFromNoteParcel().parcel(metaSiteID: 606)))
+
+        // Assert
+        waitUntil {
+            self.tabBarController.tabNavigationController(tab: .reviews).viewControllers.count == 2
+        }
+
+        XCTAssertEqual(tabBarController.selectedIndex, WooTab.reviews.visibleIndex())
+
+        let reviewsTabNavigationController = tabBarController.tabNavigationController(tab: .reviews)
+        // A ReviewDetailsViewController should be pushed
+        assertThat(reviewsTabNavigationController.topViewController, isAnInstanceOf: ReviewDetailsViewController.self)
+    }
 }
 
 private extension MainTabBarController {
     var tabRootViewControllers: [UIViewController] {
         viewControllers?.compactMap { $0 as? UINavigationController }.compactMap { $0.viewControllers.first } ?? []
     }
+
+    func tabNavigationController(tab: WooTab) -> UINavigationController {
+        guard let navigationController = viewControllers?.compactMap({ $0 as? UINavigationController })[tab.visibleIndex()] else {
+            fatalError("Unexpected access to navigation controller at tab: \(tab)")
+        }
+        return navigationController
+    }
```

Comment:
I think we should avoid `fatalError` in unit tests since they will cause crashes and halt the running tests. It does take a bit more effort to code around them though. 

```suggestion
    func tabNavigationController(tab: WooTab) -> UINavigationController? {
        viewControllers?.compactMap({ $0 as? UINavigationController })[tab.visibleIndex()]
    }
```

We'll have to unwrap all the time. 🙈 

```swift
        let reviewsTabNavigationController = try XCTUnwrap(tabBarController.tabNavigationController(tab: .reviews))
```

Maybe there's a better way somewhere. Anyway, up to you!

Reply:
Yea, it's not ideal for the tests to crash because of this scenario. I ended up just using the optional value, as in: https://github.com/woocommerce/woocommerce-ios/pull/5044/commits/8b9dc633fb1eddabfabb4f251fd3de9aa69849e5